src/core/core.datasetController.js

```diff
@@ -419,9 +415,9 @@ export default class DatasetController {
 		let sorted = true;
 		let i, parsed, cur, prev;
 
-		if (start > 0) {
+		if (start > 0 || (start === 0 && count < data.length)) {
 			sorted = meta._sorted;
-			prev = meta._parsed[start - 1];
+			prev = start > 0 ? meta._parsed[start - 1] : undefined;
 		}
```

Comment:
```suggestion
		let sorted = start === 0 && count === data.length ? true : meta._sorted;
		let prev = start > 0 && meta._parsed[start - 1];
		let i, cur, parsed;
```

Reply:
Good simplification. With fresh eyes, I thought of some cases where this method could incorrectly detect the sorted status of the data array.

## Incorrectly Mark Data as not Sorted

```
const data = [0, 2, 1, 3, 4];
parse(0, 5); // not sorted

data[1] = 1;
data[2] = 2; // data == [0, 1, 2, 3, 4]

parse(0, 3); // not sorted
```

## Incorrect Mark Data as Sorted

```javascript
const data = [0, 1, 2, 3, 4];
parse(0, 5); // sorted

data[2] = 4; // data == [0, 1, 4, 3, 4]
parse(0, 3); // sorted
```

Reply:
Good points. We should set the `meta._sorted` to false when data is modified (and parsing is enabled).
And add tests for these cases, to make sure it works :)